File: epi_judge_java_solutions/epi/MaxTrappedWater.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.List;

public class MaxTrappedWater {
  @EpiTest(testDataFile = "max_trapped_water.tsv")

  public static int getMaxTrappedWater(List<Integer> heights) {

    int i = 0, j = heights.size() - 1, maxWater = 0;
    while (i < j) {
      int width = j - i;
      maxWater =
          Math.max(maxWater, width * Math.min(heights.get(i), heights.get(j)));
      if (heights.get(i) > heights.get(j)) {
        --j;
      } else { // heights.get(i) == heights.get(j).
        ++i;
      }
    }
    return maxWater;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "MaxTrappedWater.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_cpp/max_trapped_water.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;

int GetMaxTrappedWater(const vector<int>& heights) {
    int maxWater = 0;
    int i = 0, j = heights.size() - 1;
    while (i < j) {
        maxWater = std::max(maxWater, (j - i) * std::min(heights[i], heights[j]));
        if(heights[i]>heights[j]){ 
            j--;
        }
        else if (heights[i] < heights[j]) {
            i++;
        }
        else {
            i++; j--;
        }
    }
    return maxWater;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"heights"};
  return GenericTestMain(args, "max_trapped_water.cc", "max_trapped_water.tsv",
                         &GetMaxTrappedWater, DefaultComparator{}, param_names);
}
